File: Entrega2/NumericMethodsJava/Jacobi.java
/**
 *
 * @author Valeria
 */
public class Jacobi {

    public static double[] jacobi(double[][] A, double[] b, double t, int iter, double[] x0) {
        int n = A.length;
        int l = A[0].length;
        double[][] T = new double[n][n];
        double[] C = new double[n];
        if (n != l) {
            System.out.println("A is not a square matrix.");
            return new double[0];
        } else {
            double[] x = new double[n];
            double aux;
            int cont = 0;
            double E = t + 1;
            int iteration = 1;
            while (E > t && cont <= iter) {
                System.out.println("iter: " + iteration);
                E = 0;
                for (int i = 0; i < n; i++) {
                    double sum = 0;
                    for (int j = 0; j < n; j++) {
                        if (i != j) {
                            sum = sum + A[i][j] * x0[j];
                            T[i][j] = -A[i][j] / A[i][i];
                            C[i] = b[i] / A[i][i];
                        }
                    }
                    x[i] = (b[i] - sum) / A[i][i];
                    aux = x[i] - x0[i];
                    E = E + Math.pow(aux, 2);
                }
                E = Math.pow(E, 0.5);
                System.out.println("E = " + E);
                for (int i = 0; i < n; i++) {
                    x0[i] = x[i];
                    System.out.println("x" + (i + 1) + ": " + x0[i]);

                }
                System.out.println("");
                iteration++;
                cont = cont++;
            }
            System.out.println("T: ");
            for (int i = 0; i < T.length; i++) {
                for (int j = 0; j < T.length; j++) {
                    System.out.print(T[i][j] + "      ");
                }
                System.out.println(" ");
            }
            System.out.println("C: ");
            for (int i = 0; i < T.length; i++) {
                System.out.print(C[i] + "      ");
            }
            System.out.println(" ");
            if (E < t) {
                return x;
            } else {
                System.out.println("Can not find a solution in " + iter + " iterations");
                return new double[0];
            }
        }
    }

    public static void main(String[] args) {
        double[][] A
                = {
                    {4, -1, 0, 3},
                    {1, 15.5, 3, 8},
                    {0, -1.3, -4, 1.1},
                    {14, 5, -2, 30}
                };
        double[] b = {1, 1, 1, 1};
        double[] x0 = {0, 0, 0, 0};
        jacobi(A, b, Math.pow(10, -7), 100, x0);
    }
}

File: Entrega2/NumericMethodsPython/Jacobi.py

import math
import numpy as np
#Author Valeria

def jacobi(A, b, t, iter,x0):
    n = len(A)
    l = len(A)
    if (n != l):
        print("A no es una matriz cuadrada.")
        return 0
    else:
        x = [n]
        T = []
        for i in range(n):
            T.append([0]*n)
        C = []
        aux = 0
        cont = 0
        E = t + 1
        iteration = 1
        while (E > t and cont <= iter):
            print("iter: " , iteration)
            E = 0;
            for i in range(0,n):
                suma = 0
                for j in range(0,n):
                    if (i != j):
                        suma = suma + A[i][j] * x0[j]
                        T[i][j] = -A[i][j] / A[i][i]
                C.append(b[i] / A[i][i])
                x.append((b[i] - suma) / A[i][i])
                aux = x[i] - x0[i]
                E = E + math.pow(aux, 2)
            
            E = math.pow(E, 0.5)
            print("E = " , E)
            for i in range(n):
                x0[i] = x[i]
                print("x" , (i + 1) , ": " , x0[i])
            
            print("\nT: ")
            for i in range(0,n):
                for j in range(0,n):
                    print(T[i][j] , end="      ")
                
                print(" ")
            
            print("C: ")
            for i in range(0,n):
                print(C[i] , end= "      ")
            print(" ")
            print(" ")
            
            values, normalized_eigenvectors = np.linalg.eig(T) # T es la matriz
            spectral_radius = max(abs(values))
            print("Spectral Radius: ", spectral_radius)

            iteration = iteration + 1 
            cont = cont + 1
        
        if (E < t):
            return x
        else:
            print("no se llegó a la solución en " + iter + " iteraciones");
            return 0

A = [[4, -1, 0, 3],
    [1, 15.5, 3, 8],
    [0, -1.3, -4, 1.1],
    [14, 5, -2, 30]]
b = [1, 1, 1, 1]
x0 = [0, 0, 0, 0]
t = math.pow(10, -7)
iter = 100
jacobi(A, b, t, iter, x0)

